libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -2339,6 +2339,202 @@ bool WaveTrack::Get(size_t iChannel, size_t nBuffers,
    });
 }
 
+namespace {
+void RoundToNearestClipSample(const WaveTrack& track, double& t)
+{
+   const auto clip = track.GetClipAtTime(t);
+   if (!clip)
+      return;
+   t = clip->SamplesToTime(clip->TimeToSamples(t - clip->GetPlayStartTime())) +
+       clip->GetPlayStartTime();
+}
+}
+
+std::pair<size_t, size_t> WaveTrack::GetFloatsCenteredAroundTime(
+   double t, size_t iChannel, float* buffer, size_t numSideSamples,
+   bool mayThrow) const
+{
+   const auto numSamplesReadLeft = GetFloatsFromTime(
+      t, iChannel, buffer, numSideSamples, mayThrow, PlaybackDirection::backward);
+   const auto numSamplesReadRight = GetFloatsFromTime(
+      t, iChannel, buffer + numSideSamples, numSideSamples + 1, mayThrow,
+      PlaybackDirection::forward);
+   return { numSideSamples - numSamplesReadLeft,
+            numSideSamples + numSamplesReadRight };
+}
+
+namespace
+{
+template<typename BufferType>
+struct BufferCharType;
+
+template<>
+struct BufferCharType<float*>
+{
+   using type = samplePtr;
+};
+
+template<>
+struct BufferCharType<const float*>
+{
+   using type = constSamplePtr;
+};
+
+template <typename BufferType> struct SampleAccessArgs
+{
+   const typename BufferCharType<BufferType>::type offsetBuffer;
+   const sampleCount start;
+   const size_t len;
+};
+
+template <typename BufferType>
+SampleAccessArgs<BufferType> GetSampleAccessArgs(
+   const WaveClip& clip, double startOrEndTime /*absolute*/, BufferType buffer,
+   size_t totalToRead, size_t alreadyRead, bool forward)
+{
+   const auto remainingToRead = totalToRead - alreadyRead;
+   const auto sampsInClip = clip.GetVisibleSampleCount();
+   const auto sampsPerSec = clip.GetRate() / clip.GetStretchRatio();
+   if (forward)
+   {
+      const auto startTime =
+         std::max(startOrEndTime - clip.GetPlayStartTime(), 0.);
+      const sampleCount startSamp { std::ceil(
+         std::round(startTime * sampsPerSec * 1e6) / 1e6) };
+      if (startSamp >= sampsInClip)
+         return { nullptr, sampleCount { 0u }, 0u };
+      const auto len =
+         limitSampleBufferSize(remainingToRead, sampsInClip - startSamp);
+      return { reinterpret_cast<typename BufferCharType<BufferType>::type>(
+                  buffer + alreadyRead),
+               startSamp, len };
+   }
+   else
+   {
+      const auto endTime = std::min(
+         startOrEndTime - clip.GetPlayStartTime(), clip.GetPlayDuration());
+      const sampleCount endSamp { std::ceil(
+         std::round(endTime * sampsPerSec * 1e6) / 1e6) };
+      const auto startSamp =
+         std::max(endSamp - remainingToRead, sampleCount { 0 });
+      const auto len = (endSamp - startSamp).as_size_t();
+      if (len == 0 || startSamp >= sampsInClip)
+         return { nullptr, sampleCount { 0u }, 0u };
+      const auto bufferEnd = buffer + remainingToRead;
+      return { reinterpret_cast<typename BufferCharType<BufferType>::type>(
```

Comment:
Similar question whether this is the right thing when `min` above did some work

Reply:
If `min` did some work, then `startOrEndTime` was past the end of the clip.
> should the first member of the return be a later position?

It may be confusing to have this buffer offsetting amidst this more complex code. All the offset is doing is moving the read-write pointer according to the number of samples already read/written. It should not be concerned with the number of samples remaining to read from the clip.